### ISSUE TITLE
[Bug Report]: BatchProduceAsync seems to produce items out or order

### ISSUE THREAD

**Issue by AlexeyRaga** [Farfetch/kafkaflow#594]
### Prerequisites

- [X] I have searched [issues](https://github.com/Farfetch/kafkaflow/issues) to ensure it has not already been reported

### Description

It looks like `BatchProduceAsync` can produce message out-of-order messages.
It happens from time to time, so the order doesn't seem to be deterministic.
_Most_ of the time it does seem to produce messages in the right order, but not always.

### Steps to reproduce

I was using this code to reproduce:

```csharp
private sealed record Batch(Guid Id, List<ISpecificRecord> Messages);

private static Batch CreateBatch()
{
    var id = Guid.NewGuid();
    var messages = new List<ISpecificRecord>
    {
        new UserJoined { id = id, order = 1 },
        new UserDeactivated { id = id, order = 2 },
        new UserActivated { id = id, order = 3 }
    };

    return new(id, messages);
}

[Fact]
public async Task Should_preserve_order_of_messages()
{
    var batchesToSend = Enumerable.Range(0, 10).Select(_ => CreateBatch()).ToList();

    foreach (var batch in batchesToSend)
    {
        var toProduce =
            batch.Messages
                .Select(x => new BatchProduceItem(
                    fixture.TopicName,
                    batch.Id.ToString(),
                    x,
                    new MessageHeaders()))
                .ToList();

        await fixture.Producer.BatchProduceAsync(toProduce);
    }

    // assertion is omitted for brevity 
}
```

I then look at the topic itself and see that _some_ batches come out of order.

### Expected behavior

I expect that the messages order is _guaranteed_ by the batch producer _all the time_.

### Actual behavior

Most often messages are coming in the right order: `1, 2, 3` but sometimes they are written in the wrong order.

### KafkaFlow version

v3.0.10

**Comment by AlexeyRaga** [comment]
Is it possible that the problem can be in the middleware?

I see that [BatchProduceAsync](https://github.com/Farfetch/kafkaflow/blob/master/src/KafkaFlow/Producers/BatchProduceExtension.cs#L36) executes `Produce` synchronously for each message, separately.

But it internally it does `_middlewareExecutor.Execute` which returns `Task`. This `Task` is dropped on the floor.
This could mean that multiple `Task`s are just running in parallel and by chance can finish out-of-order.

If I am right, then this execution model is unsound.

**Comment by massada** [comment]
That is exactly why is not guaranteeing in order producing. For it to guarantee in order producing, the code would need to wait until InternalProduce is called (this method pushed the message to the driver buffer).

It should ne very easy to reproduce this issue by adding a producing middleware with a descending timeout and see an inverse order of producing.

Let's say 3 messages with descending timeouts of 1s, 500ms, 0s.

**Comment by mattstermiller** [comment]
I can confirm this is an issue. My messages were produced out-of-order when using BatchProduceAsync until I removed producer serializer middleware. Order is preserved when using a middleware that calls the next one synchronously. Like @massada said, the call to InternalProduce needs to be synchronous, so the middlewares need to be executed synchronously somehow.

**Comment by AlexeyRaga** [comment]
> the middlewares need to be executed synchronously somehow

I don't think that we must guarantee that middlewares are executed synchronously, or at least I don't see what would require that.
What needs to be guaranteed that the underlaying producer is called in order.
So perhaps it should register all the exits from the middleware (final producer calls), track the list and as soon as it is in order dispatch to the final producer.
This way middlewares would be allowed to be asynchronous, but the messages would be sent in sync.

I wonder if it can be done as something like a batching middleware too...

Something like:

- a batch producer creates an `AsyncLocal` value with the list of `(message, isprocessed)` in order
- The "final" level of a middleware (of a "real" producer wrapper) checks if that `AsyncLocal` is present
  - If not - just call the real producer and be done
  - If yes, check if the message can be dispatched and dispatch accordingly... Or wait until all the messages in the list are ready and then dispatch them...

**Comment by mattstermiller** [comment]
Let me share the workaround I came up with. I don't love it, but it works. It's a pair of middlewares that sandwich other middleware and use a semaphore to serialize the calls into the middleware pipeline.

```cs
/// <summary>
/// Provides support for ordered batch producing.
/// </summary>
public static class BatchProducerSynchronizationMiddleware
{
    /// <summary>
    /// Adds middlewares at the beginning and end of the pipeline that forces parallel calls to the middleware pipeline
    /// to be executed serially to preserve ordering for batch producing. This should be called on the middleware
    /// builder last.
    /// </summary>
    /// <remarks>
    /// The only way to produce in batches is to call the inner Confluent producer multiple times without waiting for a
    /// result. The asynchronous nature of the middleware pipeline means that these calls will race to call the
    /// Confluent producer to enqueue their messages. Without this middleware, the result is indeterministic ordering.
    /// </remarks>
    public static IProducerMiddlewareConfigurationBuilder AddBatchSynchronization(
        this IProducerMiddlewareConfigurationBuilder middlewares
    )
        => middlewares.AddAtBeginning<SynchronizationBeginMiddleware>()
            .Add<SynchronizationEndMiddleware>();


    private const string SemaphoreItemName = "SynchronizationSemaphore";

    private static void ReleaseContextSemaphore(IMessageContext context)
    {
        if (context.Items[SemaphoreItemName] is not SemaphoreSlim semaphore)
        {
            return;
        }

        context.Items[SemaphoreItemName] = null;
        semaphore.Release();
    }

    internal class SynchronizationBeginMiddleware : IMessageMiddleware
    {
        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);

        public async Task Invoke(IMessageContext context, MiddlewareDelegate next)
        {
            await _semaphore.WaitAsync();
            context.Items[SemaphoreItemName] = _semaphore;
            try
            {
                await next.Invoke(context).ConfigureAwait(false);
            }
            catch (Exception)
            {
                ReleaseContextSemaphore(context);
                throw;
            }
        }
    }

    internal class SynchronizationEndMiddleware : IMessageMiddleware
    {
        public Task Invoke(IMessageContext context, MiddlewareDelegate next)
        {
            try
            {
                return next.Invoke(context);
            }
            finally
            {
                ReleaseContextSemaphore(context);
            }
        }
    }
}
```

To use it, tack on a call to `.AddBatchSynchronization()` to your producer's middleware.

**Comment by mattstermiller** [comment]
To fix the problem properly, I think the BatchProduceAsync method needs to be built into the KafkaFlow producer itself. That way, it could call the middleware pipeline for each message serially (using await) instead of in parallel, then call the inner producer after all the middlewares are done.

**Comment by mattstermiller** [comment]
It really bothers me that the behavior of the synchronous Produce method is unpredictable though. I suggested making it invoke the middleware pipeline synchronously to avoid issues like this in general.
